worker/caasoperator/caasoperator.go

```diff
@@ -291,6 +290,23 @@ func (op *caasOperator) makeAgentSymlinks(unitTag names.UnitTag) error {
 		}
 	}
 
+	// Ensure legacy charm symlinks created before 2.8 getting unlinked.
```

Comment:
I'd refactor this like to keep the success flow to the end of the function.
Makes it easier for the next person to add to the function.
```
	// Ensure legacy charm symlinks created before 2.8 getting unlinked.
	unitCharmDir := filepath.Join(op.config.DataDir, "agents", unitTag.String(), "charm")
	isUnitCharmDirSymlink, err := jujusymlink.IsSymlink(unitCharmDir)
	if os.IsNotExist(errors.Cause(err)) || os.IsPermission(errors.Cause(err)) {
		// Ignore permission denied as this won't happen in production
		// but may happen in testing depending on setup of /tmp	
	} else if err != nil {
		return errors.Trace(err)
	} else if isUnitCharmDirSymlink {
		op.config.Logger.Debugf("unlinking legacy charm symlink for unit %q", unitTag)
		if err := os.Remove(unitCharmDir); err != nil {
			return errors.Trace(err)
		}
	}
```